aten/src/ATen/native/ReduceOpsUtils.h

```diff
@@ -260,4 +257,45 @@ static TensorIterator make_reduction(
   return make_reduction(name, result1, result2, self, dim, keepdim, dtype, dtype);
 }
 
+static void zero_numel_check_dims(const Tensor& self, const int64_t dim, const char *fn_name) {
+  if (self.ndimension() == 0) {
+    TORCH_CHECK_INDEX(dim == 0 || dim == -1, fn_name,
+      ": Expected reduction dim -1 or 0 for scalar but got ", dim);
+  }
+  else {
+    TORCH_CHECK_INDEX(self.size(dim) != 0, fn_name,
+      ": Expected reduction dim ", dim, " to be non-zero.");
+  }
+}
+
+static void zero_numel_check_dims(const Tensor& self, const IntArrayRef dim, const char *fn_name) {
+  for (const int64_t d : dim) {
+    zero_numel_check_dims(self, d, fn_name);
+  }
+}
+
+// Resize the result tensor and indices when result.numel() == 0 depending on values of
+// dim and keepdim for returning tensors containing reduction results.
+// This function should be called when you are reducing a zero-dim tensor and want to
+// simply resize the output and return it.
```

Comment:
This description is not very clear. For one, it's wrong: s/zero-dim tensor/zero-numel tensor/. If there is an invariant that the passed in tensor should have zero numel, you should internal assert it inside this function.  But also it seems worth explaining here why this function exists; it exists because although normally you're not allowed to reduce on a zero numel dimension, if there are zero elements, you can do it because you won't ever be called out on the value in question.

Reply:
Works. Done.